spec/models/provider_spec.rb

```diff
@@ -34,38 +34,146 @@
+              'address4' => enrichment.address4,
+              'postcode' => enrichment.postcode,
+              'region_code' => ProviderEnrichment.region_codes[enrichment.region_code]
+            )
+          end
+
+          it 'returns json_data from the newest enrichment' do
+            enrichment = build(:provider_enrichment)
+            newest_enrichment = build(:provider_enrichment, created_at: Date.today)
+            provider = create(:provider, enrichments: [enrichment, newest_enrichment])
+
+            expect(provider.address_info).to eq(
+              'address1' => newest_enrichment.address1,
+              'address2' => newest_enrichment.address2,
+              'address3' => newest_enrichment.address3,
+              'address4' => newest_enrichment.address4,
+              'postcode' => newest_enrichment.postcode,
+              'region_code' => ProviderEnrichment.region_codes[newest_enrichment.region_code]
+            )
+          end
+        end
+        context 'enrichment has partial set for address_info' do
+          it 'returns address of the enrichment' do
+            enrichment = build(:provider_enrichment, json_data: { 'email' => Faker::Internet.email,
+            'website' => Faker::Internet.url,
+            'address1' => Faker::Address.street_address,
+            'postcode' => nil,
+            'train_with_us' => Faker::Lorem.sentence.to_s,
+            'train_with_disability' => Faker::Lorem.sentence.to_s })
+            provider = create(:provider, enrichments: [enrichment])
+
+            expect(provider.address_info).to eq(
+              'address1' => enrichment.address1,
+              'address2' => enrichment.address2,
+              'address3' => enrichment.address3,
+              'address4' => enrichment.address4,
+              'postcode' => enrichment.postcode,
+              'region_code' => ProviderEnrichment.region_codes[enrichment.region_code]
+            )
+          end
+
+          context 'enrichment has only region code set for address_info' do
+            it 'returns address of the provider' do
+              london = ProviderEnrichment.region_codes['London']
+              enrichment = build(:provider_enrichment, json_data: { 'email' => Faker::Internet.email,
+              'website' => Faker::Internet.url,
+              'region_code' => 'No region',
+              'train_with_us' => Faker::Lorem.sentence.to_s,
+              'train_with_disability' => Faker::Lorem.sentence.to_s })
+              provider = create(:provider, region_code: london, enrichments: [enrichment])
+
+              expect(provider.address_info).to eq(
+                'address1' => provider.address1,
+                'address2' => provider.address2,
+                'address3' => provider.address3,
+                'address4' => provider.address4,
+                'postcode' => provider.postcode,
+                'region_code' => ProviderEnrichment.region_codes[provider.region_code]
```

Comment:
Change `ProviderEnrichment.region_codes[provider.region_code]` to `provider.region_code_before_type_cast`.